### PR TITLE
Fixtures with Files and Images

### PR DESCRIPTION
**Pull request by bobdenotter**

### CODE REVIEW

File: src/Content/MediaFactory.php [hunk 1]
```diff
@@ -31,29 +33,33 @@ class MediaFactory
 
     /** @var Collection */
     private $mediaTypes;
+    private $areas;
 
-    public function __construct(Config $config, MediaRepository $mediaRepository, TokenStorageInterface $tokenStorage)
+    public function __construct(Config $config, Areas $areas, MediaRepository $mediaRepository, TokenStorageInterface $tokenStorage)
```
Comment (JarJak) [inline src/Content/MediaFactory.php]: what is `Areas` class for?

File: src/DataFixtures/ImagesFixtures.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    private $urls = [];
+
+    public const AMOUNT = 10;
+
+    /** @var MediaFactory */
+    private $mediaFactory;
+
+    public function __construct(Config $config, Areas $areas, MediaFactory $mediaFactory)
+    {
+        $this->urls = [
```
Comment (JarJak) [inline src/DataFixtures/ImagesFixtures.php]: save as Collection

File: src/DataFixtures/ImagesFixtures.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    private function fetchImages(): void
+    {
+        $outputPath = $this->areas->get('files', 'basepath') . '/stock/';
+
+        if (! is_dir($outputPath)) {
+            mkdir($outputPath);
+        }
+
+        for ($i = 1; $i <= $this::AMOUNT; $i++) {
```
Comment (JarJak) [inline src/DataFixtures/ImagesFixtures.php]: self::AMOUNT

File: src/DataFixtures/ImagesFixtures.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            $url = $this->urls[array_rand($this->urls)] . random_int(10000, 99999);
```
Comment (JarJak) [inline src/DataFixtures/ImagesFixtures.php]: $this->urls->random()

File: src/DataFixtures/ImagesFixtures.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+use Faker\Factory;
+use Faker\Generator;
+use GuzzleHttp\Client;
+
+class ImagesFixtures extends BaseFixture
+{
+    /** @var Generator */
+    private $faker;
+
+    private $urls = [];
```
Comment (JarJak) [inline src/DataFixtures/ImagesFixtures.php]: no need to set default value

File: src/DataFixtures/ImagesFixtures.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    public const AMOUNT = 10;
```
Comment (JarJak) [inline src/DataFixtures/ImagesFixtures.php]: private

File: src/DataFixtures/BaseFixture.php [hunk 7]
```diff
@@ -4,13 +4,30 @@
...
+    /** @var Config */
+    protected $config;
+
+    /** @var Areas */
+    protected $areas;
+
+    public function __construct(Config $config, Areas $areas)
+    {
+        $this->config = $config;
+        $this->areas = $areas;
```
Comment (JarJak) [inline src/DataFixtures/BaseFixture.php]: Why this constructor is added here when none of these dependencies is used in this class? 

Areas should not be a part of BaseFixture. move it to trait or to another abstract class

File: src/DataFixtures/ContentFixtures.php [hunk 8]
```diff
@@ -22,19 +23,20 @@ class ContentFixtures extends BaseFixture implements DependentFixtureInterface
...
     private $lastTitle = null;
 
     /** @var array */
     private $presetRecords = [];
 
-    public function __construct(Config $config)
+    /** @var Collection */
+    private $imagesIndex;
+
+    public function __construct(Config $config, Areas $areas)
```
Comment (JarJak) [inline src/DataFixtures/ContentFixtures.php]: changed in constructor here can be reverted, just add Areas

File: src/Entity/Media.php [hunk 9]
```diff
@@ -281,8 +281,10 @@ public function getAuthor(): User
         return $this->author;
     }
 
-    public function setAuthor(?User $author): void
+    public function setAuthor(?User $author): self
     {
         $this->author = $author;
+
+        return $this;
```
Comment (JarJak) [inline src/Entity/Media.php]: do we use chain of setters somewhere?

File: templates/_partials/fields/image.html.twig [hunk 10]
```diff
@@ -6,7 +6,7 @@
     <editor-image
         :name="'{{ name }}'"
         :filename="'{{ field.get('filename') }}'"
-        :thumbnail="'{{ field|thumbnail(width=1200, height=1200, area='files', path=setPath) }}'"
+        :thumbnail="'{{ field|thumbnail(width=1200, height=1200, area='files') }}'"
```
Comment (JarJak) [inline templates/_partials/fields/image.html.twig]: area param does nothing during thumbnail generation, is this something you plan to add later?

File: src/Content/MediaFactory.php (hunk 1 above)
Comment (bobdenotter) [reply src/Content/MediaFactory.php]: Currently there's `config`, `files` and `themes`. It allows uploading / reading and listing files in those areas. Right now locations are fixed, but we might want to make them configurable.

File: src/Entity/Media.php (hunk 9 above)
Comment (bobdenotter) [reply src/Entity/Media.php]: we do now :-)

```
            $media->setAuthor($this->getRandomReference('user'))
                ->setDescription($this->faker->paragraphs(3, true))
                ->setCopyright('© Unsplash');
```

File: src/DataFixtures/ImagesFixtures.php (hunk 6 above)
Comment (bobdenotter) [reply src/DataFixtures/ImagesFixtures.php]: done

File: src/DataFixtures/ImagesFixtures.php (hunk 5 above)
Comment (bobdenotter) [reply src/DataFixtures/ImagesFixtures.php]: done

File: src/DataFixtures/ImagesFixtures.php (hunk 4 above)
Comment (bobdenotter) [reply src/DataFixtures/ImagesFixtures.php]: done

File: src/DataFixtures/ImagesFixtures.php (hunk 3 above)
Comment (bobdenotter) [reply src/DataFixtures/ImagesFixtures.php]: done

File: src/DataFixtures/ImagesFixtures.php (hunk 2 above)
Comment (bobdenotter) [reply src/DataFixtures/ImagesFixtures.php]: done

File: templates/_partials/fields/image.html.twig (hunk 10 above)
Comment (bobdenotter) [reply templates/_partials/fields/image.html.twig]: Maybe, but if we do, i can put it back.. removed for now.

File: src/Content/MediaFactory.php (hunk 1 above)
Comment (JarJak) [reply src/Content/MediaFactory.php]: Can we think of a better name for it? `FilesLocations` maybe?

File: src/Content/MediaFactory.php [hunk 11]
```diff
@@ -31,29 +33,33 @@ class MediaFactory
...
         $this->tokenStorage = $tokenStorage;
 
         $this->exif = Reader::factory(Reader::TYPE_NATIVE);
         $this->mediaTypes = $config->getMediaTypes();
+        $this->areas = $areas;
     }
 
     public function createOrUpdateMedia(SplFileInfo $file, string $area, ?string $title = null): Media
     {
+        $path = Path::makeRelative($file->getPath(). '/', $this->areas->get($area, 'basepath'));
```
Comment (JarJak) [inline src/Content/MediaFactory.php]: $this->filesLocations->getLocation($location)->getBasepath();

**Pull request merged**